File: cryptocurrencyCore/src/main/java/com/outsourcebooster/cryptocurrency/core/rest/ValueController.java
package com.outsourcebooster.cryptocurrency.core.rest;

import com.outsourcebooster.cryptocurrency.core.model.Currency;
import com.outsourcebooster.cryptocurrency.core.model.Value;
import com.outsourcebooster.cryptocurrency.core.repository.ValueRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import java.lang.reflect.Field;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by rklimemnko on 26.05.2016.
 */
@RestController
@CrossOrigin({"http://localhost:8081","http://localhost:4200"})
@RequestMapping("/value")
public class ValueController {

    @Autowired
    private ValueRepository valueRepository;

    @RequestMapping("/all")
    public List<Value> getDataAll() {
        return valueRepository.findAll();
    }

    @RequestMapping("/minute")
    public List<Value> getDataMinute() {
        return valueRepository.findByDateBetween(LocalDateTime.now().minusMinutes(1l), LocalDateTime.now());
    }

    @RequestMapping("/hour")
    public List<Value> getDataHour() {
        return valueRepository.findByDateBetween(LocalDateTime.now().minusHours(1l), LocalDateTime.now());
    }

    @RequestMapping("/day")
    public List<Value> getDataDay() {
        return valueRepository.findByDateBetween(LocalDateTime.now().minusDays(1l), LocalDateTime.now());
    }

    @RequestMapping("/week")
    public List<Value> getDataWeek() {
        return valueRepository.findByDateBetween(LocalDateTime.now().minusWeeks(1l), LocalDateTime.now());
    }

    @RequestMapping("/month")
    public List<Value> getDataMonth() {
        return valueRepository.findByDateBetween(LocalDateTime.now().minusMonths(1l), LocalDateTime.now());
    }


    @RequestMapping("/year")
    public List<Value> getDataYear() {
        return valueRepository.findByDateBetween(LocalDateTime.now().minusYears(1l), LocalDateTime.now());
    }

    @RequestMapping("/last")
    public Value getLast() {
        return valueRepository.findFirstByOrderByDateDesc();
    }

    @RequestMapping("/currencies")
    public List<String> getCurrencies() {
        List<String> currencies = new ArrayList<>();
        for(Field field : Value.class.getDeclaredFields()) {
            if(field.getType() == Currency.class) {
                currencies.add(field.getName());
            }
        }
        return currencies;
    }
}

File: cryptocurrencyCore/src/main/java/com/outsourcebooster/cryptocurrency/core/rest/ValueController.kt
package com.outsourcebooster.cryptocurrency.core.rest;

import com.outsourcebooster.cryptocurrency.common.model.Currency
import com.outsourcebooster.cryptocurrency.common.model.Value
import com.outsourcebooster.cryptocurrency.core.repository.ValueRepository;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import java.time.LocalDateTime;

/**
 * Created by rklimemnko on 26.05.2016.
 */
@RestController
@CrossOrigin(*arrayOf("http://localhost:8081","http://localhost:4200"))
@RequestMapping("/value")
open class ValueController(val valueRepository: ValueRepository) {

    @RequestMapping("/all")
    fun getDataAll() = valueRepository.findAll()

    @RequestMapping("/minute")
    fun getDataMinute() = valueRepository.findByDateBetween(LocalDateTime.now().minusMinutes(1L), LocalDateTime.now())

    @RequestMapping("/hour")
    fun getDataHour() = valueRepository.findByDateBetween(LocalDateTime.now().minusHours(1L), LocalDateTime.now())

    @RequestMapping("/day")
    fun getDataDay() = valueRepository.findByDateBetween(LocalDateTime.now().minusDays(1L), LocalDateTime.now())

    @RequestMapping("/week")
    fun getDataWeek() = valueRepository.findByDateBetween(LocalDateTime.now().minusWeeks(1L), LocalDateTime.now())

    @RequestMapping("/month")
    fun getDataMonth() = valueRepository.findByDateBetween(LocalDateTime.now().minusMonths(1L), LocalDateTime.now())


    @RequestMapping("/year")
    fun getDataYear() = valueRepository.findByDateBetween(LocalDateTime.now().minusYears(1L), LocalDateTime.now())

    @RequestMapping("/last")
    fun getLast() = valueRepository.findFirstByOrderByDateDesc()

    @RequestMapping("/currencies")
    fun getCurrencies(): List<String> {
        val currencies = mutableListOf<String>();
        for(field in Value::class.java.getDeclaredFields()) {
            if(field.getType() == Currency::class.java) {
                currencies.add(field.getName());
            }
        }
        return currencies.toList();
    }
}
